### PR TITLE
feat: allows to define a prefix for collections / tables

### LINKED ISSUE #4715: [repository] Allows to define a prefix for collections / tables
It could be interesting to be able to define prefixes to share a database between multiple components (APIM / AM).

Edit 2021-02-24

It could be interesting for the reviewer to test both JDBC and MOngo implementations.

Here are some topics to insist on the tests because there were more complex to migrate :

- Dictionnaries
- Identity Provider
- Pages
- Rate Limit
- Quality rules

### PR DESCRIPTION
**Pull request by ytvnr**
Fixes gravitee-io/issues#4715

- [x] Test container
  - [x] postgresql-tc
  - [x] mysql-tc
  - [x] mariadb-tc
  - [x] sqlserver-tc
- [ ] Test embedded
   - [x] postgresql-te  
   - [x] mysql-te
   - [ ] mariadb-te bug -> https://github.com/gravitee-io/issues/issues/5125

### CODE REVIEW

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcAlertRepository.java [hunk 1]
```diff
@@ -41,26 +42,29 @@
...
-            .addColumn("definition", Types.NVARCHAR, String.class)
-            .addColumn("created_at", Types.TIMESTAMP, Date.class)
-            .addColumn("updated_at", Types.TIMESTAMP, Date.class)
-            .addColumn("template", Types.BIT, boolean.class)
-            .build();
+    private static String SELECT_ESCAPED_ALERT_TRIGGERS_TABLE_NAME;
+
+    JdbcAlertRepository(@Value("${management.jdbc.prefix:}") String tablePrefix) {
+        super(tablePrefix, "alert_triggers");
+        SELECT_ESCAPED_ALERT_TRIGGERS_TABLE_NAME = "select * from " + escapeReservedWord(this.tableName);
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcAlertRepository.java]: Does 'alert_triggers' need to be escaped ?

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java [hunk 2]
```diff
@@ -124,8 +127,8 @@ private boolean addClause(boolean first, StringBuilder query) {
     public Set<ApiKey> findBySubscription(String subscription) throws TechnicalException {
         LOGGER.debug("JdbcApiKeyRepository.findBySubscription({})", subscription);
         try {
-            List<ApiKey> apiKeys = jdbcTemplate.query("select * from " + escapeReservedWord("keys") + " where subscription = ?"
-                    , ORM.getRowMapper()
+            List<ApiKey> apiKeys = jdbcTemplate.query("select * from " + escapeReservedWord(this.tableName) + " where subscription = ?"
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java]: escape the table name in the constructor
```suggestion
            List<ApiKey> apiKeys = jdbcTemplate.query(getOrm().getSelectAllSql() + " where subscription = ?"

```

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java [hunk 3]
```diff
@@ -139,7 +142,7 @@ private boolean addClause(boolean first, StringBuilder query) {
     public Set<ApiKey> findByPlan(String plan) throws TechnicalException {
         LOGGER.debug("JdbcApiKeyRepository.findByPlan({})", plan);
         try {
-            List<ApiKey> items = jdbcTemplate.query("select * from " + escapeReservedWord("keys") + " where plan = ?"
+            List<ApiKey> items = jdbcTemplate.query("select * from " + escapeReservedWord(this.tableName) + " where plan = ?"
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java]: escape the table name in the constructor
```suggestion
            List<ApiKey> items = jdbcTemplate.query(getOrm().getSelectAllSql() + " where plan = ?"

```

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcCommandRepository.java [hunk 4]
```diff
@@ -94,10 +92,10 @@ protected String getId(Command item) {
     public Optional<Command> findById(String id) throws TechnicalException {
         LOGGER.debug("JdbcCommandRepository.findById({})", id);
         try {
-            JdbcHelper.CollatingRowMapper<Command> rowMapper = new JdbcHelper.CollatingRowMapper<>(ORM.getRowMapper(), CHILD_ADDER, "id");
-            jdbcTemplate.query("select * from " + escapeReservedWord("commands") + " c " +
-                            "left join command_acknowledgments ca on c.id = ca.command_id " +
-                            "left join command_tags ct on c.id = ct.command_id " +
+            JdbcHelper.CollatingRowMapper<Command> rowMapper = new JdbcHelper.CollatingRowMapper<>(getOrm().getRowMapper(), CHILD_ADDER, "id");
+            jdbcTemplate.query("select * from " + escapeReservedWord(this.tableName) + " c " +
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcCommandRepository.java]: escape table name in the constructor
```suggestion
            jdbcTemplate.query(getOrm().getSelectAllSql() + " c " +
```

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcCommandRepository.java [hunk 5]
```diff
@@ -159,16 +157,16 @@ public Command update(Command item) throws TechnicalException {
     @Override
     public List<Command> search(CommandCriteria criteria) {
         LOGGER.debug("JdbcCommandRepository.search({})", criteria);
-        JdbcHelper.CollatingRowMapper<Command> rowMapper = new JdbcHelper.CollatingRowMapper<>(ORM.getRowMapper(), CHILD_ADDER, "id");
+        JdbcHelper.CollatingRowMapper<Command> rowMapper = new JdbcHelper.CollatingRowMapper<>(getOrm().getRowMapper(), CHILD_ADDER, "id");
         final StringBuilder query = new StringBuilder(
-                "select * from " + escapeReservedWord("commands") + " c " +
-                "left join command_acknowledgments ca on c.id = ca.command_id " +
-                "left join command_tags ct on c.id = ct.command_id " +
+                "select * from " + escapeReservedWord(this.tableName) + " c " +
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcCommandRepository.java]: escape the table name in the constructor
```suggestion
                getOrm().getSelectAll() + " c " +
```

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcParameterRepository.java [hunk 6]
```diff
@@ -123,8 +132,8 @@ public void delete(String key, String referenceId, ParameterReferenceType refere
     public Optional<Parameter> findById(String key, String referenceId, ParameterReferenceType referenceType) throws TechnicalException {
         LOGGER.debug("JdbcParameterRepository.findById({}, {}, {})", key, referenceId, referenceType);
         try {
-            final List<Parameter> items = jdbcTemplate.query("select * from parameters where " + escapeReservedWord("key") + " = ? and reference_type = ? and reference_id = ?"
-                    , ORM.getRowMapper()
+            final List<Parameter> items = jdbcTemplate.query(getOrm().getSelectAllSql() + " where " + escapeReservedWord("key") + " = ? and reference_type = ? and reference_id = ?"
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcParameterRepository.java:135]: getSelectAll() escapes the table name.
```suggestion
            final List<Parameter> items = jdbcTemplate.query("select * from " + this.tableName + " where " + escapeReservedWord("key") + " = ? and reference_type = ? and reference_id = ?"

```

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcParameterRepository.java [hunk 7]
```diff
@@ -144,14 +153,14 @@ public void delete(String key, String referenceId, ParameterReferenceType refere
             if (isEmpty(keys)) {
                 return Collections.emptyList();
             }
-            List<Parameter> parameters = jdbcTemplate.query("select * from parameters where reference_id = ? and reference_type = ? and " + escapeReservedWord("key") + " in ( "
-                            + ORM.buildInClause(keys) + " )"
+            List<Parameter> parameters = jdbcTemplate.query(getOrm().getSelectAllSql() + " where reference_id = ? and reference_type = ? and " + escapeReservedWord("key") + " in ( "
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcParameterRepository.java:156]: getSelectAll() escapes the table name.

```suggestion
            List<Parameter> parameters = jdbcTemplate.query("select * from " + this.tableName + " where reference_id = ? and reference_type = ? and " + escapeReservedWord("key") + " in ( "

```

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcParameterRepository.java [hunk 8]
```diff
@@ -164,8 +173,8 @@ public void delete(String key, String referenceId, ParameterReferenceType refere
     public List<Parameter> findAll(String referenceId, ParameterReferenceType referenceType) throws TechnicalException {
         LOGGER.debug("JdbcParameterRepository.findAll({}, {})", referenceId, referenceType);
         try {
-            List<Parameter> parameters = jdbcTemplate.query("select * from parameters where reference_id = ? and reference_type = ?"
-                    , ORM.getRowMapper()
+            List<Parameter> parameters = jdbcTemplate.query(getOrm().getSelectAllSql() + " where reference_id = ? and reference_type = ?"
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/management/JdbcParameterRepository.java:176]: getSelectAll() escapes the table name.

```suggestion
            List<Parameter> parameters = jdbcTemplate.query("select  * from " + this.tableName + " where reference_id = ? and reference_type = ?"

```

File: src/main/java/io/gravitee/repository/jdbc/ratelimit/JdbcRateLimitRepository.java [hunk 9]
```diff
@@ -49,10 +50,13 @@
...
 
+    @Value("${management.jdbc.prefix:}")
+    private String prefix;
+
     private final TransactionTemplate transactionTemplate;
 
-    private static String buildInsertStatement() {
-        return "insert into ratelimit (" +
+    private String buildInsertStatement() {
+        return "insert into " + getTableNameFor("ratelimit") + " (" +
```
Comment (phiz71) [inline src/main/java/io/gravitee/repository/jdbc/ratelimit/JdbcRateLimitRepository.java]: Add a constructor and initialize an instance variable with the table name and the prefix

**Review by phiz71 — CHANGES_REQUESTED** [review]
On top of the review comments, I think you should replace all `getTableNameFor("xxx")` with instance variables, initialized in the constructor.

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java (hunk 2 above)
Comment (ytvnr) [reply src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java]: done

File: src/main/java/io/gravitee/repository/jdbc/management/JdbcAlertRepository.java (hunk 1 above)
Comment (ytvnr) [reply src/main/java/io/gravitee/repository/jdbc/management/JdbcAlertRepository.java]: i used getSelectAll()

**Review by phiz71 — APPROVED** [review]

**Pull request merged**